contracts/governance/GraphGovernance.sol

```diff
@@ -0,0 +1,92 @@
+// SPDX-License-Identifier: MIT
+
+pragma solidity ^0.7.3;
+
+import "../upgrades/GraphUpgradeable.sol";
+
+import "./GraphGovernanceStorage.sol";
+
+/**
+ * @title Graph Governance Contract
+ * @notice Governance contract used to inscribe Graph Council and community votes.
+ */
+contract GraphGovernance is GraphGovernanceV1Storage, GraphUpgradeable, IGraphGovernance {
+    // -- Events --
+
+    event ProposalCreated(
+        address submitter,
+        bytes32 proposalId,
+        bytes32 votes,
+        bytes32 metadata,
+        ProposalResolution resolution
+    );
+    event ProposalUpdated(
+        address submitter,
+        bytes32 proposalId,
+        bytes32 votes,
+        bytes32 metadata,
+        ProposalResolution resolution
+    );
+
+    /**
+     * @notice Initialize this contract.
+     */
+    function initialize(address _governor) public onlyImpl {
+        Governed._initialize(_governor);
+    }
+
+    // -- Proposals --
+
+    /**
+     * @notice Return whether the proposal is created.
+     * @param _proposalId Proposal identifier
+     * @return True if the proposal is already created
+     */
+    function isProposalCreated(bytes32 _proposalId) public view override returns (bool) {
+        return proposals[_proposalId].votes != 0;
+    }
+
+    /**
+     * @notice Submit a new proposal.
+     * @param _proposalId Proposal identifier. This is an IPFS hash to the content of the proposal
+     * @param _votes An IPFS hash of the collection of signatures for each vote
```

Comment:
a thought - brandon mentioned in our meeting yesterday that not all proposals will necessarily have a Snapshot vote (i.e. IPFS hash) associated with them. 

What I understand right now is that `createProposal` will always be linked to a  Graph Governance Proposal (GGP), and a GGP may or may not be related to a GIP, which may or may not be associated with a Snapshot IPFS hash

Ariel, wdyt? How do you understand it right now?

Reply:
also, a GGP can be created before or after a contract upgrade. in an emergency, the upgrade would happen before the GGP was sent out to the community

Reply:
Now i am thinking out loud, maybe this is how it works:
- a GGP is done on snapshot. This will always happen for an upgrade. Sometimes before, sometimes after. But it does not matter if it is before or after, as we are just storing governance proposals on chain.
- Therefore the proposal ID is the GGP ipfs hash
- Therefore the community votes, and the GIP ipfs hash, are not being uploaded here. If anything, the GGP will reference any necessary GIPs, and at GIP will mention any necessary snapshot votes 

WDYT ??

Reply:
> a GGP is done on snapshot. This will always happen for an upgrade. Sometimes before, sometimes after. But it does not matter if it is before or after, as we are just storing governance proposals on chain.

Sorry for the noise, but in the mid-term this could happen in Defender, so the process of 1) setting up the tx, 2) collecting all council signatures, and 3) publishing it to IPFS, happens all within the same app.

Reply:
This seems ideal to me. For now we can get the council to work inside of both snapshot and defender. but in the future if defender could use a snapshot plugin that created proposals as well as voted on proposals, that would be great. it would make the council members tasks easier to handle